### DISCUSSION TITLE
Is it possible to render [[this-file]] link syntax used in my markdown files?

### DISCUSSION THREAD

**Discussion by HiRolla** [joomlatools/joomlatools-pages#875] [Ideas]
Hello! As I mentioned in another question, I have Obsidian markdown files. I'm going to publish them via Joomlatools-Pages, add commenting system, etc.  And now my question is about linking files via [[file-name]] syntax. 

I can see that Pages deals with (classic)[md-link] syntax pretty cool. I think that I can find some syntax converter, but first I would like to ask if there some other workarounds via Pages?

If so, then there is also an issue with the ommited path in my [[file-name-links]].  Does Pages have any mechanism or scripting flexibility to fullfill ommitted path when rendering [[file-name-links]]?

**Comment by johanjanssens** [comment]
Simple answer no. As far as I know `[[...]]` isn't standard Markdown or CommonMark syntax. Github uses it for relative links to Pages, some other Markdown extensions seem to use it for other purposes.

Pages makes use of [PHP Markdown](https://michelf.ca/projects/php-markdown/) for rendering markdown files. It offers a [`url_filter_func`](https://michelf.ca/projects/php-markdown/configuration/) to handle url routing but not sure if that supports a `[[ ... ]]`syntax. 

Do you know about any resources describing how `[[... ]]`is intended to work in Markdown?

**Comment by HiRolla** [comment]
> some other Markdown extensions seem to use it for other purposes

Thanks! I did not knew that. 

> isn't standard Markdown or CommonMark syntax

Yes, now I see that originally this is wikilink syntax, not markdown. I've met this syntax in [Obsidian.md](https://help.obsidian.md/How+to/Internal+link) and Wikipedia.org. Here is description https://en.wikipedia.org/wiki/Help:Link#Wikilinks_(internal_links) of this syntax.

In short, this is `[[target|label]]` syntax, where `text ` part is optional.

Also I've seen almost the similar syntax in TiddlyWiki, but there parts are swapped [[label|target]], And it seems to be a deviation.

So we really can't say there is some kind of such [[linking]] standard. Also there are some project-specific extentions of this syntax like `[[target#chapter|label]]` in Obsidian or `[[cheese]]s` in Wikipedia where label = target + extension ["s"], no pipe needed.

Thus it seems more reasonable to convert all links to the markdown standard: `[label](target)` before placing them into joomlatools/pages or /data. I think this is not to hard to do using regexp in the source .md files. Another way is to convert links on the fly just before before rendering them (I'm not sure). The second approach is more interesting to me but the first one is more clear to me now.

**Reply by johanjanssens** [reply]
I'm most happy to try to add this in some way. Seems we need an Obsedian flavoured parser or parser plugin for this. I found following:

- https://forum.obsidian.md/t/is-there-a-parser-renderer-reference-spec/29504/2
- https://help.obsidian.md/How+to/Format+your+notes

The [[ ... ]] syntax is described as internal links here. I suspect you want to be able to use it to link to other pages easily, this would be a markdown'ish implementation for the `route([page])` template function. 

The next question then is how do we extend the markdown parser, might be better to make the parser pluggable and allow using diff parsers; i do have a related ticket here: https://github.com/joomlatools/joomlatools-pages/issues/378

**Reply by HiRolla** [reply]
Yes, it would be just great to be able to link via `[[ ... ]]` syntax! Thank you for going to deal with it! Also it's a great idea to make pluggable parsers!

I've read that Obsidian topics too. But I see one important thing that is not described there. The thing is that links like this [[ ... ]] in Obsidian may not include the path, just a filename without extention. Maybe there is no need to follow this approach because of variant readings. Isn't it?

**Reply by johanjanssens** [reply]
It's pretty limited that way indeed. Think for your use-case, it would be a specific plugin that supports the Obisidian syntax, this would allow you to seemingly use your Obsidian md files with Pages, worth a try.

**Reply by HiRolla** [reply]
> a specific plugin that supports the Obisidian syntax

Hope this going to happen! I love Joomla, Pages, Obsidian and Syncthing. Make them work together — this is what my dreams are about!
I would like to learn anythyng is needed for that. @johanjanssens Thank you for paying attention on my questions and for making such a great products 💎

**Reply by johanjanssens** [reply]
You are most welcome. That sounds like a great combo indeed, i'm a bit too busy with other code right now to spend time on this but I have moved this discussion to the ideas category. Can pick it up later again.
